### PR TITLE
lib/ukrust: Update ukrust to work alongside Cargo

### PR DESCRIPTION
**Pull request by vladandrew**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -325,6 +325,8 @@ GOCFLAGS :=
 GOCFLAGS-y :=
 RUSTCFLAGS :=
 RUSTCFLAGS-y :=
+CARGOFLAGS :=
+CARGOFLAGS-y :=
```
Comment (skuenzer) [inline Makefile]: Do we really need cargo in the core repository? I'd prefer to have it part of an external language support library.

Comment (vladandrew) [reply Makefile]: Hi Simon, I'll actually move all the build rules to the external library.  This PR is a late stage draft.

File: lib/ukrust/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "ukrust"
+version = "0.1.0"
+edition = "2018"
```
Comment (TrueDoctor) [inline lib/ukrust/Cargo.toml:5]: We should probably also add a license here

**Comment by fabianpatras** [comment]
This PR fixes the typo found in #596.

File: lib/ukrust/Cargo.toml (hunk 2 above)
Comment (vladandrew) [reply lib/ukrust/Cargo.toml:5]: Makes sense!

File: lib/nolibc/string.c [hunk 3]
```diff
@@ -815,3 +815,8 @@ char *strncat(char *dest, const char *src, size_t n)
 	*dest++ = 0;
 	return a;
 }
+
+int bcmp(const void *s1, const void *s2, size_t count)
```
Comment (skuenzer) [inline lib/nolibc/string.c:819]: As far as I saw, `bcmp` is an obsolete libc function and `memcmp` should be used instead. Where is `bcmp` still used?

File: support/build/Makefile.build [hunk 4]
```diff
@@ -53,9 +55,11 @@ $(foreach S,$($(call uc,$(L))_SRCS) $($(call uc,$(L))_SRCS-y) \
 	    $(EACHOLIB_SRCS) $(EACHOLIB_SRCS-y), \
 $(eval $(call buildrule_libobj_multitarget,$(L),$(S))) \
 ); \
-$(if $(call qstrip,$($(call uc,$(L))_OBJS) $($(call uc,$(L))_OBJS-y)), \
+$(if $(call qstrip,$($(call uc,$(L))_OBJS) $($(call uc,$(L))_OBJS-y) \
+                   $($(call uc,$(L))_ALIBS) $($(call uc,$(L))_ALIBS-y)), \
 $(eval $(call buildrule_olib,$(L))); \
 $(eval $(call uc,$(P))_OLIBS-y    += $(call libname2olib,$(L))); \
+$(eval UK_ALIBS-y += $($(call uc,$(L))_ALIBS) $($(call uc,$(L))_ALIBS-y)); \
```
Comment (skuenzer) [inline support/build/Makefile.build]: Hum... I am not sure if this makes so much sense. If you want to link with a static library at the last step, you would register such a lib to `UK_ALIBS-y` in your `Makefile.uk` (like with `UK_ALIBS-$(CONFIG_MYLIB)+=mylib.a`). Similar to other scopes, it was intended that static libraries that are registered to a library focus are linked before the `exportsyms.uk` is applied. How useful this feature is, i a different question because linking with static libraries work differently (simple form of DCE).

File: Makefile [hunk 5]
```diff
@@ -588,6 +588,11 @@ RUSTC		:= rustc --target=$(CONFIG_LLVM_TARGET_ARCH)
 else
 RUSTC		:= rustc
 endif
+ifneq ("$(origin LLVM_TARGET_ARCH)","undefined")
+CARGO		:= cargo --target=$(CONFIG_LLVM_TARGET_ARCH)
```
Comment (skuenzer) [inline Makefile]: Hum 🤔  Where is this used? ...and for which reasons do we need cargo in the core repo?

Comment (vladandrew) [reply Makefile]: The user should be able to set his cargo path this way. Something like `make CARGO=xcargo or /path/to/cargo`.

File: support/build/Makefile.build (hunk 4 above)
Comment (vladandrew) [reply support/build/Makefile.build:62]: `UK_ALIBS-$(CONFIG_MYLIB)+=mylib.a)` doing this right now does not work as expected. The alib is imidiatly link to an object file.

File: lib/nolibc/string.c (hunk 3 above)
Comment (vladandrew) [reply lib/nolibc/string.c:819]: rustc generate code that uses bcmp. LLVM 9.0 still emits calls to `bcmp`[1]. I was getting undefined symbol errors when building with cargo.

[1] https://lists.llvm.org/pipermail/llvm-bugs/2019-March/072889.html

File: support/build/Makefile.build (hunk 4 above)
Comment (vladandrew) [reply support/build/Makefile.build]: Actually you're right, this exact line doesn't make sense. I've removed these changes from the pull request. The only thing remaining is to link alibs at the end, not before the exportsyms.uk is applied.

File: Makefile (hunk 5 above)
Comment (skuenzer) [reply Makefile]: I mean, I do not see a build rule using `$(CARGO)`. The thing is, if cargo is used by a external support lib, you should implement this with this library's `Makefile.rules`. We could also add support to load external `Makefile.build`. @marcrittinghaus looked into this.

File: lib/nolibc/string.c (hunk 3 above)
Comment (skuenzer) [reply lib/nolibc/string.c:819]: Hum, okay. Then I wonder two points:
1) Could we introduce an `alias` to `memcmp` instead of defining a new function?
2) I am not yet sure what the best location for the alias definition would be. Do you know if `musl` provides `bcmp`? If yes, it has to be `nolibc`. If not, we may want to place the alias with the rust support library that uses `cargo`.

Ah... and maybe there is a third option where we could handover a built-in option for `bcmp` to LLVM calls?

File: support/build/Makefile.rules [hunk 6]
```diff
@@ -873,10 +873,6 @@ $(call libname2preolib,$(1)): $$($(call vprefix_lib,$(1),OBJS)) \
 			      $$($(call vprefix_lib,$(1),OBJS-y)) \
 			      $$(EACHOLIB_OBJS) \
 			      $$(EACHOLIB_OBJS-y) \
-			      $$($(call vprefix_lib,$(1),ALIBS)) \
-			      $$($(call vprefix_lib,$(1),ALIBS-y)) \
-			      $$(EACHOLIB_ALIBS) \
-			      $$(EACHOLIB_ALIBS-y) \
```
Comment (skuenzer) [inline support/build/Makefile.rules]: Hum... 🤔 either this commits needs a different description why we remove the ALIB option for libraries or the patch should not be part of the PR?

File: Makefile (hunk 5 above)
Comment (vladandrew) [reply Makefile]: I've moved it to the library's `Makefile.rules`[1] :confetti_ball: .

[1] https://github.com/unikraft/lib-rust/blob/staging/Makefile.rules#L39

File: lib/nolibc/string.c (hunk 3 above)
Comment (vladandrew) [reply lib/nolibc/string.c:819]: Yes, musl 1.1.19 (the version ported on unikraft) defines bcmp at `src/string/bcmp.c`. They use the same call to `memcpy`, so I'd keep this as a function and not an alias to match musl's approach. I wouldn't complicate things with the handover.

File: support/build/Makefile.rules (hunk 6 above)
Comment (vladandrew) [reply support/build/Makefile.rules]: Got it, yeah, I've removed this commit entirely

File: lib/ukrust/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+[package]
+license-file = "../../LICENSE.txt"
```
Comment (mkroening) [inline lib/ukrust/Cargo.toml]: Which license file does this reference? Preferably, we would specify `license` instead of `license-file`: https://doc.rust-lang.org/cargo/reference/manifest.html#the-license-and-license-file-fields

File: lib/ukrust/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+name = "ukrust"
+version = "0.1.0"
+edition = "2018"
```
Comment (TrueDoctor) [inline lib/ukrust/Cargo.toml:5]: This should be updated to 2021 right?

File: lib/ukrust/ukrust_sys/lib.rs [hunk 9]
```diff
@@ -49,3 +49,15 @@ pub mod c_types;
...
+extern "C" {
+    fn __ukrust_sys_crash() -> !;
+}
+
+#[panic_handler]
+fn panic(_info: &core::panic::PanicInfo<'_>) -> ! {
+    unsafe {
+        __ukrust_sys_crash();
+    }
+}
```
Comment (fabianpatras) [inline lib/ukrust/ukrust_sys/lib.rs:62]: line 54: ```warning: `__ukrust_sys_crash` redeclared with a different signature``` 
It is saying about `__ukrust_sys_crash` exported by `bindings`.

Can we remove the explicit ` -> !` to make the warning go away and insert a `loop {}` in the panic handler?

Like the examples from https://doc.rust-lang.org/nomicon/panic-handler.html

File: lib/ukrust/Cargo.toml (hunk 7 above)
Comment (vladandrew) [reply lib/ukrust/Cargo.toml]: Good catch! I think unikraft is under the BSD 3 license, but not sure if there are changes. This is the main license file https://github.com/unikraft/unikraft/blob/staging/COPYING.md , I'll update the path to it.

File: lib/ukrust/Cargo.toml (hunk 8 above)
Comment (vladandrew) [reply lib/ukrust/Cargo.toml:5]: Yeah, I did try that initially, but it doesn't build because the core crate is built using the 2018 edition[1].  Though, in the future we should definitely update it. Thanks Dennis!

[1] https://github.com/unikraft/unikraft/blob/staging/lib/ukrust/Makefile.uk#L19

File: lib/ukrust/ukrust_sys/lib.rs (hunk 9 above)
Comment (vladandrew) [reply lib/ukrust/ukrust_sys/lib.rs:62]: Hi Fabian, I would keep it like this right now, since this would be a fix, and this pull request is related only to moving. Maybe add it in a separate pull request, but I remember they didn't use the `loop {}` in Linux neither, so I wonder why.

File: lib/ukrust/Cargo.toml (hunk 8 above)
Comment (TrueDoctor) [reply lib/ukrust/Cargo.toml:5]: Ah, that makes sense

File: lib/ukrust/ukrust_sys/lib.rs (hunk 9 above)
Comment (fabianpatras) [reply lib/ukrust/ukrust_sys/lib.rs:62]: Hi Vlad,

I see Linux using `loop {}` this way inside the `panic_handler`. [1]
This is, however, a workaround for a reason mentioned there ([1]) in the comments.

I understand your reasoning. Let's add it next time then.

[1] - https://github.com/torvalds/linux/blob/041fae9c105ae342a4245cf1e0dc56a23fbb9d3c/rust/kernel/lib.rs#L78-L87

Comment (vladandrew) [reply lib/ukrust/ukrust_sys/lib.rs:62]: Cool, they updated it! Yeah, we should do an upgrade in the future.

**Review by fabianpatras — COMMENTED** [review]
Looks good to me (this I can only say about the Rust part). :ok_hand: 
There are things to modify, but as you mentioned (in the comments), they are not the scope of this PR and they should be done in a separate PR.

I don't have "Approve" or "Request changes" rights so I can only "Comment".

Reviewed-by: Fabian Patras <patras.fabi@gmail.com>

**Comment by razvand** [comment]
@fabianpatras , you can hit the approve button now, I added you as a reviewer.

**Review by fabianpatras — APPROVED** [review]

File: lib/ukrust/ukrust_sys/lib.rs [hunk 10]
```diff
@@ -49,3 +49,15 @@ pub mod c_types;
...
+
```
Comment (razvand) [inline lib/ukrust/ukrust_sys/lib.rs]: I would remove the extra blank line.

**Review by razvand — CHANGES_REQUESTED** [review]
Thanks, @vladandrew , this works great. I would update the commit message to:

```
    To make life easier for Cargo, Rust crates should add `ukrust`
    as a dependency. We update `ukrust` to include a `Cargo.toml` file and
    to match the expected structure of a Cargo project. We use `src/` as a symbolic link directory
    pointing to `ukrust_sys/`.
```

File: lib/ukrust/ukrust_sys/lib.rs (hunk 10 above)
Comment (vladandrew) [reply lib/ukrust/ukrust_sys/lib.rs]: Done

**Review by razvand — APPROVED** [review]
Thanks, @vladandrew 

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
